File: src/com/kiibou/server/GameState.java
package com.kiibou.server;

import com.kiibou.TileType;
import space.kiibou.data.Vec2;
import space.kiibou.data.Vec3;

import java.util.*;
import java.util.stream.Collectors;

import static com.kiibou.TileType.*;

public class GameState {
    private final GameService gameService;
    private long handle;
    private int width;
    private int height;
    private int bombs;
    private boolean[][] revealed;
    private boolean[][] flagged;
    private TileType[][] tiles;
    private Vec2[] bombTiles;
    private boolean gameRunning;
    private int revealedTiles;
    private Timer timer;
    private TimerTask timerTask;
    private int time;

    public GameState(final long handle, final int width, final int height, final int bombs, final GameService gameService) {
        this.handle = handle;
        setupVariables(width, height, bombs);
        timer = new Timer("Timer", true);
        this.gameService = Objects.requireNonNull(gameService);
    }

    void setupVariables() {
        setupVariables(width, height, bombs);
    }

    private void setupVariables(final int width, final int height, final int bombs) {
        this.width = width;
        this.height = height;
        this.bombs = bombs;

        if (gameRunning) {
            stopTimer();
        }

        revealed = new boolean[width][height];
        for (boolean[] value : revealed) {
            Arrays.fill(value, false);
        }

        flagged = new boolean[width][height];
        for (boolean[] booleans : flagged) {
            Arrays.fill(booleans, false);
        }

        tiles = new TileType[width][height];
        for (TileType[] tile : tiles) {
            Arrays.fill(tile, EMPTY);
        }

        // Find out where to put bombs
        final List<Vec2> emptyTiles = new ArrayList<>();
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                emptyTiles.add(new Vec2(x, y));
            }
        }
        Collections.shuffle(emptyTiles);
        bombTiles = emptyTiles.subList(0, bombs).toArray(new Vec2[0]);
        for (Vec2 coordinate : bombTiles) {
            tiles[coordinate.x][coordinate.y] = BOMB;
        }

        // Set tiles around bombs to the correct number
        for (int x = 0; x < tiles.length; x++) {
            for (int y = 0; y < tiles[x].length; y++) {
                if (isBomb(x, y)) continue;
                int count = 0;

                for (int px = -1; px <= 1; px++) {
                    for (int py = -1; py <= 1; py++) {
                        if (isValidTile(x + px, y + py) && isBomb(x + px, y + py)) {
                            count++;
                        }
                    }
                }

                setTile(x, y, getTypeFromValue(count));
            }
        }

        gameRunning = false;
        revealedTiles = 0;
    }

    List<Vec3> reveal(final int x, final int y) {
        if (!gameRunning) setGameRunning(true);

        final List<Vec3> revealed = new ArrayList<>();

        if (isValidTile(x, y)) {
            if (getTile(x, y) == EMPTY) {
                for (int px = -1; px <= 1; px++) {
                    for (int py = -1; py <= 1; py++) {
                        if (revealTile(x + px, y + py, revealed)) {
                            revealed.addAll(reveal(x + px, y + py));
                        }
                    }
                }
            } else {
                if (isBomb(x, y)) {
                    setTile(x, y, RED_BOMB);
                    setGameRunning(false);
                    revealed.addAll(
                            Arrays.stream(bombTiles)
                                    .filter(vec -> !(vec.x == x && vec.y == y))
                                    .map(vec -> new Vec3(vec.x, vec.y, BOMB.getLookup()))
                                    .collect(Collectors.toList())
                    );

                    gameService.sendLoose(handle);
                    setGameRunning(false);
                }

                revealTile(x, y, revealed);
            }
        }

        if (revealedTiles == width * height - bombs && gameRunning) {
            gameService.sendWin(handle);
            Vec2[] notFlagged = Arrays.stream(bombTiles)
                    .filter(tile -> !isFlagged(tile.x, tile.y))
                    .toArray(Vec2[]::new);

            for (Vec2 tile : notFlagged) {
                gameService.sendFlagToggle(handle, tile.x, tile.y);
            }

            setGameRunning(false);
        }

        return revealed;
    }

    private boolean revealTile(final int x, final int y, final List<Vec3> revealed) {
        boolean val = false;

        if (isValidTile(x, y)) {
            if (isNotRevealed(x, y)) {
                if (getTile(x, y) == EMPTY) val = true;
                revealed.add(new Vec3(x, y, getTile(x, y).getLookup()));
                setRevealed(x, y, true);
                revealedTiles++;
            }
        }

        return val;
    }

    private TileType getTile(final int x, final int y) {
        return tiles[x][y];
    }

    private void setTile(final int x, final int y, final TileType tileType) {
        Objects.requireNonNull(tileType);

        tiles[x][y] = tileType;
    }

    private boolean isNotRevealed(final int x, final int y) {
        return !revealed[x][y];
    }

    private void setRevealed(final int x, final int y, final boolean r) {
        if (isFlagged(x, y)) {
            gameService.sendFlagToggle(handle, x, y);
        }

        revealed[x][y] = r;
    }

    private boolean isFlagged(final int x, final int y) {
        return flagged[x][y];
    }

    private boolean isBomb(final int x, final int y) {
        return getTile(x, y) == BOMB;
    }

    private boolean isValidTile(final int x, final int y) {
        return x >= 0 && x < width && y >= 0 && y < height;
    }

    private void setGameRunning(final boolean running) {
        if (!gameRunning && running) {
            resetTimer();
            startTimer();
        } else if (gameRunning && !running) {
            stopTimer();
        }

        gameRunning = running;
    }

    boolean flagToggle(final int x, final int y) {
        if (!gameRunning) setGameRunning(true);
        if (isNotRevealed(x, y)) {
            flagged[x][y] = !flagged[x][y];
        }
        return flagged[x][y];
    }

    private void startTimer() {
        timerTask = new TimerTask() {
            @Override
            public void run() {
                sendTimeToClients();
                time++;
            }
        };
        timer.schedule(timerTask, 0, 1000);
    }

    private void stopTimer() {
        if (timerTask != null)
            timerTask.cancel();
    }

    private void resetTimer() {
        time = 0;
        sendTimeToClients();
    }

    private void sendTimeToClients() {
        gameService.sendTime(handle, time);
    }

}

File: src/com/kiibou/server/GameState.kt
package com.kiibou.server

import com.kiibou.TileInfo
import com.kiibou.TileType
import space.kiibou.data.Vec2
import java.util.*
import kotlin.streams.toList

class GameState(private val handle: Long, width: Int, height: Int, bombs: Int, private val gameService: GameService) {
    private var width = 0
    private var height = 0
    private var bombs = 0
    private lateinit var revealed: Array<BooleanArray>
    private lateinit var flagged: Array<BooleanArray>
    private lateinit var tiles: Array<Array<TileType>>
    private lateinit var bombTiles: Array<Vec2>
    private var gameRunning = false
    private var revealedTiles = 0
    private val timer = Timer("Timer", true)
    private lateinit var timerTask: TimerTask
    private var time = 0

    fun setupVariables() = setupVariables(width, height, bombs)

    fun setupVariables(width: Int, height: Int, bombs: Int) {
        this.width = width
        this.height = height
        this.bombs = bombs

        if (gameRunning) {
            stopTimer()
        }

        resetTimer()

        revealed = Array(width) { BooleanArray(height) { false } }
        flagged = Array(width) { BooleanArray(height) { false } }
        tiles = Array(width) { Array(height) { TileType.EMPTY } }

        // Find out where to put bombs
        bombTiles = placeBombs()

        // Set tiles around bombs to the correct number
        createNumberTiles()

        gameRunning = false
        revealedTiles = 0
    }

    private fun possibleTilePositions(x: Int = 0, y: Int = 0, width: Int = this.width, height: Int = this.height) =
            List(width * height) { Vec2(x + it % width, y + it / height) }

    private fun placeBombs() = chooseBombPositions().run(::setTilesToBombs)
    private fun chooseBombPositions() = possibleTilePositions().shuffled().take(bombs)
    private fun setTilesToBombs(list: List<Vec2>) = list.onEach { (x, y) -> setTile(x, y, TileType.BOMB) }

    private fun createNumberTiles() = possibleTilePositions()
            .filterNot { (x, y) -> isBomb(x, y) }
            .associateWith { (x, y) -> countSurroundingBombs(x, y) }
            .forEach { (pos, count) -> setTile(pos.x, pos.y, TileType.getTypeFromValue(count)) }

    private fun countSurroundingBombs(x: Int, y: Int) = possibleTilePositions(-1, -1, 3, 3)
            .filter { (px, py) -> isValidTile(x + px, y + py) && isBomb(x + px, y + py) }.count()

    fun reveal(x: Int, y: Int): List<TileInfo> {
        if (!gameRunning) setGameRunning(true)
        val revealed: MutableList<TileInfo> = ArrayList()

        if (isValidTile(x, y)) {
            when (getTile(x, y)) {
                TileType.EMPTY -> possibleTilePositions(x - 1, y - 1, 3, 3)
                        .filter { (tx, ty) -> revealTile(tx, ty, revealed) }
                        .forEach { (tx, ty) -> revealed += reveal(tx, ty) }
                TileType.BOMB -> {
                    setTile(x, y, TileType.RED_BOMB)
                    setGameRunning(false)

                    revealed += bombTiles.filterNot { it == Vec2(x, y) }
                            .map { (x, y) -> TileInfo(x, y, TileType.BOMB.lookup) }

                    revealTile(x, y, revealed)
                    gameService.sendLoose(handle)
                    setGameRunning(false)
                }
                else -> revealTile(x, y, revealed)
            }
        }

        if (revealedTiles == width * height - bombs && gameRunning) {
            gameService.sendWin(handle)

            bombTiles.filter { (x, y) -> !isFlagged(x, y) }
                    .onEach { (x, y) -> flagToggle(x, y) }
                    .forEach { (x, y) -> gameService.sendFlagStatus(handle, x, y) }

            setGameRunning(false)
        }

        return revealed
    }

    private fun revealTile(x: Int, y: Int, revealed: MutableList<TileInfo>): Boolean {
        var empty = false

        if (isValidTile(x, y) && isNotRevealed(x, y)) {
            empty = getTile(x, y) == TileType.EMPTY
            revealed.add(TileInfo(x, y, getTile(x, y).lookup))
            setRevealed(x, y, true)
            revealedTiles++
        }

        return empty
    }

    private fun getTile(x: Int, y: Int) = tiles[x][y]

    private fun setTile(x: Int, y: Int, tileType: TileType) {
        tiles[x][y] = tileType
    }

    private fun isNotRevealed(x: Int, y: Int) = !revealed[x][y]

    private fun setRevealed(x: Int, y: Int, r: Boolean) {
        if (isFlagged(x, y)) {
            flagToggle(x, y)
            gameService.sendFlagStatus(handle, x, y)
        }
        revealed[x][y] = r
    }

    fun isFlagged(x: Int, y: Int) = flagged[x][y]
    private fun isBomb(x: Int, y: Int) = getTile(x, y) == TileType.BOMB
    private fun isValidTile(x: Int, y: Int) = x in 0 until width && y >= 0 && y < height

    private fun setGameRunning(running: Boolean) {
        when {
            !gameRunning && running -> {
                resetTimer()
                startTimer()
            }
            gameRunning && !running -> {
                stopTimer()
            }
        }
        gameRunning = running
    }

    fun flagToggle(x: Int, y: Int): Boolean {
        if (!gameRunning) setGameRunning(true)
        if (isNotRevealed(x, y)) {
            flagged[x][y] = !flagged[x][y]
        }
        return flagged[x][y]
    }

    private fun startTimer() {
        timerTask = object : TimerTask() {
            override fun run() {
                sendTimeToClients()
                time++
            }
        }
        timer.schedule(timerTask, 0, 1000)
    }

    private fun stopTimer() = timerTask.cancel()

    private fun resetTimer() {
        time = 0
        sendTimeToClients()
    }

    private fun sendTimeToClients() = gameService.sendTime(handle, time)

    init {
        setupVariables(width, height, bombs)
    }
}